Make some of the ifm3d modules optional

graugans opened the pull request:
The ``pcicclient`` for example is turned off by default. So do ensure only
examples are build for which the dependency is fulfilled.

:warning: Please take care before this Pull-Request can be merged the Pull-Request [#98](https://github.com/lovepark/ifm3d/pull/98) needs to be merged.

Signed-off-by: Christian Ege <christian.ege@ifm.com>

Commit:
Make some of the ifm3d modules optional

The pcicclient for example is turned off by default. So do ensure only
examples are build for which the dependency is fulfilled.

Signed-off-by: Christian Ege <christian.ege@ifm.com>

tpanzarella commented:
I have not tested this, but I don't see an obvious problem with it except scalability as the number of examples grow. I still suggest a general restructuring of the project wherein each example is its own CMake project. I think for beginners (the intention of this repo) this, as implemented, will get confusing in a different way.

I assume you (@graugans) will handle the merge of this? Also, 0.11.0 is still on a dev branch in ifm3d -- i.e., it is not yet "released".

tpanzarella commented:
UPDATE: ifm3d 0.11.0 has been pushed out as a release. Additionally, 0.12.0 has been bootstrapped -- assuming next working branch will be a minor (i.e., feature-based) update and not a patch release.

graugans commented:
Thank you very much @tpanzarella I'll update this PR so 0.11.0 is mandatory. I totally agree the way this PR handles the different modules feels kind of hacky.

So based on my limited knowledge about CMake the move into sub folders only makes things easier when we omit the top-level CMakeLists.txt. If we still want to have a top-level CMakeLists.txt we will always have to check the version and module for each sub-folder in the top-level CMakeList.txt.

Because if one of the sub folder bails out because anything is missing the whole Project will not build. Personally I like the idea of having each example as a separate and independent CMake project. I do not fear the complexity in the top-level CMakeLists.txt because we and not the users have to deal with it. So please if you have any idea how we can ensure that each CMakeLists.txt of each example bails out when building standalone if anything is missing. But on the other hand will not interrupt the whole examples when building I would love to hear your advise then.

Commit:
Make ifm3d v0.11.0 mandatory

This task depends on a fix which was released in the v0.11.0 of ifm3d.
Which allows us to make modules of ifm3d optional.

Signed-off-by: Christian Ege <christian.ege@ifm.com>

tpanzarella commented:
> So please if you have any idea how we can ensure that each CMakeLists.txt of each example bails out when building standalone if anything is missing. But on the other hand will not interrupt the whole examples when building I would love to hear your advise then.

I think there are several options, and the correct option is dependent upon your goals for the `ifm3d-examples` project.

One option is to not include a top-level `CMakeLists.txt` and literally treat each example as an independent project. This has the *downside* of having to build each subdirectory independently, but I am not so sure that is bad. First, if I am looking to learn a particular "thing" within `ifm3d`, I likely only care about one particular example, not all of them. Second, I thought one of the objectives of this project was to provide a *soup-to-nuts* example showing how easy it is to interface to `ifm3d`, including setting up your build system (we assume CMake). To that end, having each directory as a separate/independent CMake project would be the correct approach. This would be instructive to new-comers (our target audience) as it would show them how to get started from step 0.

Another option, would be to have the top-level `CMakeLists.txt` deal with the messiness of calling all the `find_package(...)` stuff, and building a list of subdirectories (examples) which are supported on the system. Then it simply loops over the list calling `add_subdirectory` as appropriate (i.e., we never call into the examples that we know will fail due to deps).

Maybe a more elegant solution is to leverage the [return command](https://cmake.org/cmake/help/v3.5/command/return.html). I don't have specific guidance on this, as I would want to prototype something first, but, it is likely worth investigation if having a top-level `CMakeLists.txt` driver script is something that you want.

If you are looking for opinions, I like "option one." It is most apropos for the goals of this poject: self-contained examples showing one thing and one thing only, including how to write a build file, aimed at beginners.

graugans commented:
I do agree with option one. I think most people will take only specific examples as a starter. If there is a big need for option 2. We can add this heavy lifting in a top-level CMake. 

The ``return`` command looks a bit hacky to me. In case of the stand-alone aproach this might lead to a succesful ``cmake ..`` without any output.
